logbus/solvermon/vertexmon.go

```diff
@@ -37,111 +37,147 @@ type vertexMonitor struct {
 	isCanceled     bool
 }
 
-var reErrExitCode = regexp.MustCompile(`^(?:process ".*" did not complete successfully|error calling LocalhostExec): exit code: (?P<exit_code>[0-9]+)$`)
-var reErrNotFound = regexp.MustCompile(`^failed to calculate checksum of ref ([^ ]*): (.*)$`)
-var reHint = regexp.MustCompile(`^(?P<msg>.+?):Hint: .+`)
+var reErrExitCode = regexp.MustCompile(`(?:process ".*" did not complete successfully|error calling LocalhostExec): exit code: (?P<exit_code>[0-9]+)$`)
 
-func (vm *vertexMonitor) Write(dt []byte, ts time.Time, stream int) (int, error) {
-	if stream == BuildkitStatsStream {
-		stats, err := vm.ssp.Parse(dt)
+func getExitCode(errString string) uint64 {
+	if reErrExitCode.MatchString(errString) {
+		matches, _ := stringutil.NamedGroupMatches(errString, reErrExitCode)
+		exitCodeMatch := matches["exit_code"][0]
+		exitCode, err := strconv.ParseUint(exitCodeMatch, 10, 32)
 		if err != nil {
-			return 0, errors.Wrap(err, "failed decoding stats stream")
+			return 0
 		}
-		for _, statsSample := range stats {
-			statsJSON, err := json.Marshal(statsSample)
-			if err != nil {
-				return 0, errors.Wrap(err, "stats json encode failed")
-			}
-			_, err = vm.cp.Write(statsJSON, ts, int32(stream))
-			if err != nil {
-				return 0, errors.Wrap(err, "write stats")
-			}
+		return exitCode
+	}
+	return 0
+}
+
+var reErrNotFound = regexp.MustCompile(`^failed to calculate checksum of ref ([^ ]*): (.*)$`)
+var reHint = regexp.MustCompile(`^(?P<msg>.+?):Hint: .+`)
+
+func determineFatalErrorType(errString string, exitCode uint64) (logstream.FailureType, bool) {
+	if strings.Contains(errString, "context canceled") || errString == "no active sessions" {
+		return logstream.FailureType_FAILURE_TYPE_UNKNOWN, false
+	}
+	if reErrExitCode.MatchString(errString) {
+		switch exitCode {
+		case math.MaxUint32:
+			return logstream.FailureType_FAILURE_TYPE_OOM_KILLED, true
+		default:
+			return logstream.FailureType_FAILURE_TYPE_NONZERO_EXIT, true
 		}
-		return len(dt), nil
 	}
-	_, err := vm.cp.Write(dt, ts, int32(stream))
-	if err != nil {
-		return 0, errors.Wrap(err, "write log line")
+	if reErrNotFound.MatchString(errString) {
+		return logstream.FailureType_FAILURE_TYPE_FILE_NOT_FOUND, true
 	}
-	return len(dt), nil
+	if strings.Contains(errString, errutil.EarthlyGitStdErrMagicString) {
+		return logstream.FailureType_FAILURE_TYPE_GIT, true
+	}
+	return logstream.FailureType_FAILURE_TYPE_UNKNOWN, false
 }
 
-func (vm *vertexMonitor) parseError() {
-	errString := vm.vertex.Error
-	indentOp := strings.Join(strings.Split(vm.operation, "\n"), "\n          ")
-	internalStr := ""
-	if vm.meta.Internal {
-		internalStr = " internal"
-	}
+func formatErrorMessage(errString, operation, internal string, fatalErrorType logstream.FailureType, exitCode uint64) string {
```

Comment:
can we change `internal` to be a bool? That way it's more clear what values we can expect `internal` to be.

then inside the function we could create a

```
internalStr := ""
if internal {
    internalStr = " internal"
}
```